File: intermediate/final/prog/digitexp.java
import java.io.*;
import java.util.*;

// Can use BigInteger to allow even bigger numbers!!!

public class digitexp {

	public static long operand(String s) {
		if(s == null || s.isEmpty() || s.charAt(0) == '0')
			return Long.MIN_VALUE;
		return Long.parseLong(s);
	}

	public static void main(String[] args) {
		int M = args.length > 0 ? Integer.parseInt(args[0]) : 10;
		Scanner input = new Scanner(System.in);
		for(int m = 0; m < M && input.hasNextLine(); ++m) {
			String digits = input.nextLine();
			long N = digits.length();
			long res = 0;
			// 1-operator
			for(int n = 1; n < N && N-n > 0; ++n) {
				String str1 = digits.substring(0,n);
				String str2 = digits.substring(n);
				long num1 = operand(str1);
				long num2 = operand(str2);
				if(num1 != Long.MIN_VALUE && num2 != Long.MIN_VALUE) {
					//cout << num1 << ' ' << num2 << endl;
					if(num1 + num2 > 0) ++res;
					if(num1 - num2 > 0) ++res;
					if(num1 * num2 > 0) ++res;
				}
			}
			// 2-operator
			for(int i = 1; i < N; ++i) {
				for(int j = 1; j < N && N-i-j > 0; ++j) {
					String str1 = digits.substring(0,i);
					String str2 = digits.substring(i,i+j);
					String str3 = digits.substring(i+j);
					long num1 = operand(str1);
					long num2 = operand(str2);
					long num3 = operand(str3);
					if(num1 != Long.MIN_VALUE && num2 != Long.MIN_VALUE && num3 != Long.MIN_VALUE) {
						//cout << num1 << ' ' << num2 << ' ' << num3 << endl;
						if(num1 + num2 - num3 > 0) ++res;
						if(num1 - num2 + num3 > 0) ++res;
						if(num1 + num2 * num3 > 0) ++res;
						if(num1 * num2 + num3 > 0) ++res;
						if(num1 - num2 * num3 > 0) ++res;
						if(num1 * num2 - num3 > 0) ++res;
					}
				}
			}
			// 3-operator
			for(int i = 1; i < N; ++i) {
				for(int j = 1; j < N; ++j) {
					for(int k = 1; k < N && N-i-j-k > 0; ++k) {
						String str1 = digits.substring(0,i);
						String str2 = digits.substring(i,i+j);
						String str3 = digits.substring(i+j,i+j+k);
						String str4 = digits.substring(i+j+k);
						long num1 = operand(str1);
						long num2 = operand(str2);
						long num3 = operand(str3);
						long num4 = operand(str4);
						if(num1 != Long.MIN_VALUE && num2 != Long.MIN_VALUE && num3 != Long.MIN_VALUE && num4 != Long.MIN_VALUE) {
							//cout << num1 << ' ' << num2 << ' ' << num3 << ' ' << num4 << endl;
							if(num1 + num2 - num3 * num4 > 0) ++res;
							if(num1 + num2 * num3 - num4 > 0) ++res;
							if(num1 * num2 - num3 + num4 > 0) ++res;
							if(num1 * num2 + num3 - num4 > 0) ++res;
							if(num1 - num2 + num3 * num4 > 0) ++res;
							if(num1 - num2 * num3 + num4 > 0) ++res;
						}
					}
				}
			}
			System.out.println(res);
		}
	}
}

File: 2018-19/finals/prog/digitexp.py
import os
import sys

def operand(s):
	if not s or len(s) == 0 or s[0] == '0':
		return sys.maxsize
	return int(s)

if __name__=="__main__":
	M = 10 if len(sys.argv) < 2 else int(sys.argv[1])
	for m in range(M):
		digits = input()
		N = len(digits)
		res = 0
		# 1-operator
		n = 1
		while n < N and N-n > 0:
			str1 = digits[0:n]
			str2 = digits[n:]
			#print(str1 + ' ' + str2)
			num1 = operand(str1)
			num2 = operand(str2)
			if num1 != sys.maxsize and num2 != sys.maxsize:
				if num1 + num2 > 0:
					res += 1
				if num1 - num2 > 0:
					res += 1
				if num1 * num2 > 0:
					res += 1
			n += 1
		# 2-operator
		for i in range(1,N):
			for j in range(1, N):
				if N-i-j <= 0:
					continue
				str1 = digits[0:i]
				str2 = digits[i:i+j]
				str3 = digits[i+j:]
				#print(str1 + ' ' + str2 + ' ' + str3)
				num1 = operand(str1)
				num2 = operand(str2)
				num3 = operand(str3)
				if num1 != sys.maxsize and num2 != sys.maxsize and num3 != sys.maxsize:
					if num1 + num2 - num3 > 0:
						res += 1
					if num1 - num2 + num3 > 0:
						res += 1
					if num1 + num2 * num3 > 0:
						res += 1
					if num1 * num2 + num3 > 0:
						res += 1
					if num1 - num2 * num3 > 0:
						res += 1
					if num1 * num2 - num3 > 0:
						res += 1
		# 3-operator
		for i in range(N):
			for j in range(N):
				for k in range(N):
					if N-i-j-k <= 0:
						continue
					str1 = digits[0:i]
					str2 = digits[i:i+j]
					str3 = digits[i+j:i+j+k]
					str4 = digits[i+j+k:]
					#print(str1 + ' ' + str2 + ' ' + str3 + ' ' + str4)
					num1 = operand(str1)
					num2 = operand(str2)
					num3 = operand(str3)
					num4 = operand(str4)
					if num1 != sys.maxsize and num2 != sys.maxsize and num3 != sys.maxsize and num4 != sys.maxsize:
						if num1 + num2 - num3 * num4 > 0:
							res += 1
						if num1 + num2 * num3 - num4 > 0:
							res += 1
						if num1 * num2 - num3 + num4 > 0:
							res += 1
						if num1 * num2 + num3 - num4 > 0:
							res += 1
						if num1 - num2 + num3 * num4 > 0:
							res += 1
						if num1 - num2 * num3 + num4 > 0:
							res += 1
		print(res)
